lib/libconfigcpp.c++

```diff
@@ -126,182 +173,214 @@ static int __toTypeCode(Setting::Type type)
 {
 }
 
 // ---------------------------------------------------------------------------
 
-const char *SettingTypeException::what() const LIBCONFIGXX_NOEXCEPT
+SettingTypeException::SettingTypeException(const Setting &setting)
+  : SettingException("setting type exception", setting)
 {
-  return("SettingTypeException");
 }
 
 // ---------------------------------------------------------------------------
 
-SettingNotFoundException::SettingNotFoundException(const Setting &setting,
-                                                   int idx)
-  : SettingException(setting, idx)
+SettingTypeException::SettingTypeException(const Setting &setting, int idx)
+  : SettingException("setting type exception", setting, idx)
 {
 }
 
 // ---------------------------------------------------------------------------
 
-SettingNotFoundException::SettingNotFoundException(const Setting &setting,
-                                                   const char *name)
-  : SettingException(setting, name)
+SettingTypeException::SettingTypeException(const Setting &setting,
+                                           const char *name)
+  : SettingException("setting type exception", setting, name)
 {
 }
 
 // ---------------------------------------------------------------------------
 
-SettingNotFoundException::SettingNotFoundException(const char *path)
-  : SettingException(path)
+SettingNotFoundException::SettingNotFoundException(const Setting &setting,
+                                                   int idx)
+  : SettingException("setting not found exception", setting, idx)
 {
 }
 
 // ---------------------------------------------------------------------------
 
-const char *SettingNotFoundException::what() const LIBCONFIGXX_NOEXCEPT
+SettingNotFoundException::SettingNotFoundException(const Setting &setting,
+                                                   const char *name)
+  : SettingException("setting not found exception", setting, name)
 {
-  return("SettingNotFoundException");
 }
 
 // ---------------------------------------------------------------------------
 
-SettingNameException::SettingNameException(const Setting &setting,
-                                           const char *name)
-  : SettingException(setting, name)
+SettingNotFoundException::SettingNotFoundException(const char *path)
+  : SettingException("setting not found exception", path)
 {
 }
 
 // ---------------------------------------------------------------------------
 
-const char *SettingNameException::what() const LIBCONFIGXX_NOEXCEPT
+SettingNameException::SettingNameException(const Setting &setting,
+                                           const char *name)
+  : SettingException("setting name exception", setting, name)
 {
-  return("SettingNameException");
 }
 
 // ---------------------------------------------------------------------------
 
-const char *FileIOException::what() const LIBCONFIGXX_NOEXCEPT
+FileIOException::FileIOException()
+  : ConfigException("file io exception")
```

Comment:
I/O
Prefer if the words were all capitalized in these descriptions

Reply:
Should I write it as one word in camel case then? As an alternative I could also turn this into a more readable description:
```suggestion
  : ConfigException("File input/output error occurred.")
```
I think this would be more consistent with the error messages from C. Additionally, for consistency, I would change the following prefixes:
```
"parse exception:" -> "Failed parsing:"
"setting not found exception" -> "Setting not found:"
"setting name exception" -> "Failed adding setting:" (Maybe the exception itself should be renamed to, e.g., SettingAddException.)

```
In addition, `SettingTypeException` seems to be used in a lot of places with different context. Maybe we can add some more context information to the exception (e.g., `Expected an array type`).

Reply:
Sure, that sounds good.

Reply:
I did not do the more precise error messages for `SettingTypeException` but I hope I addressed everything else. :)